compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/inlineclasses/InlineClassAbi.kt

```diff
@@ -0,0 +1,121 @@
+     * contain a field.
+     */
+    fun getUnderlyingType(irClass: IrClass): IrType {
+        require(irClass.isInline)
+        val primaryConstructor = irClass.constructors.single { it.isPrimary }
+        return primaryConstructor.valueParameters[0].type
+    }
+
+    /**
+     * Returns a mangled name for a function taking inline class arguments
+     * to avoid clashes between overloaded methods.
+     */
+    fun mangledNameFor(irFunction: IrFunction): Name {
+        val base = when {
+            irFunction is IrConstructor ->
+                "constructor"
+            irFunction.isGetter ->
+                JvmAbi.getterName(irFunction.propertyName.asString())
+            irFunction.isSetter ->
+                JvmAbi.setterName(irFunction.propertyName.asString())
+            irFunction.name.isSpecial ->
+                error("Unhandled special name in mangledNameFor: ${irFunction.name}")
+            else ->
+                irFunction.name.asString()
+        }
+
+        val suffix = when {
+            irFunction.fullValueParameterList.any { it.type.erasedUpperBound.isInline } ->
+                hashSuffix(irFunction)
+            (irFunction.parent as? IrClass)?.isInline == true -> "impl"
+            else -> return irFunction.name
+        }
+
+        return Name.identifier("$base-$suffix")
+    }
+
+    private val IrFunction.propertyName: Name
+        get() = (this as IrSimpleFunction).correspondingPropertySymbol!!.owner.name
+
+    private fun hashSuffix(irFunction: IrFunction) =
+        md5base64(irFunction.fullValueParameterList.joinToString { it.type.eraseToString() })
+
+    private fun IrType.eraseToString() = buildString {
+        append('L')
+        append(erasedUpperBound.fqNameWhenAvailable!!)
+        if (isNullable()) append('?')
+        append(';')
+    }
+}
+
+internal val IrFunction.hasInlineClassParameters: Boolean
+    get() = explicitParameters.any { it.type.erasedUpperBound.isInline } || (this is IrConstructor && constructedClass.isInline)
+
+internal val IrFunction.hasStaticReplacement: Boolean
+    get() = origin != IrDeclarationOrigin.FAKE_OVERRIDE &&
+            (this is IrSimpleFunction || this is IrConstructor && constructedClass.isInline && !isPrimary)
+
+internal val IrFunction.hasMethodReplacement: Boolean
+    get() = dispatchReceiverParameter != null && this is IrSimpleFunction && (parent as? IrClass)?.isInline != true
+
+internal val IrFunction.fullValueParameterList: List<IrValueParameter>
+    get() = listOfNotNull(extensionReceiverParameter) + valueParameters
+
+internal val IrClass.inlineClassFieldName: Name
+    get() = constructors.single { it.isPrimary }.valueParameters.single().name
+
+val IrFunction.isInlineClassFieldGetter: Boolean
+    get() = (parent as? IrClass)?.isInline == true && this is IrSimpleFunction &&
+            correspondingPropertySymbol?.let { it.owner.getter == this && it.owner.name == parentAsClass.inlineClassFieldName } == true
```

Comment:
It's not enough to check property name, consider:
```
inline class Z(val s: String) {
    val Int.s: Int get() = 42
}
```

(upon fixing this, please add a test if no other test becomes unmuted)

Reply:
Nice catch - this code was ported from the current backend and this is indeed a bug even in the current backend. Specifically, the code fails with a verification error.

I'll submit an issue for this and add a test.

Reply:
I've opened an issue on [YouTrack](https://youtrack.jetbrains.com/issue/KT-31927) and added a corresponding [test](https://github.com/sfs/kotlin/blob/148590b45fde3eb5a74064974a8a43e713afaa24/compiler/testData/codegen/box/inlineClasses/fieldNameClash.kt).

Unfortunately, while I've fixed this in the IR backend, there seems to be something wrong in the type mapper. Specifically, calling `mapType` on class `Z` in your example above results in an incorrect type. I'll need to look into the type mapper code a bit more before fixing this.

Reply:
OK, thanks! I suppose it can be fixed separately from this PR then.